File: firstrun/gpt-5.3codex/gpt-5.3-codex/crypto_003_naive.py
import secrets
import hashlib
import hmac
from datetime import datetime, timedelta, timezone
from typing import Dict, Any, Optional

# In-memory session store for demo purposes.
# In production, use Redis/DB and secure transport/storage.
_SESSION_STORE: Dict[str, Dict[str, Any]] = {}


def generate_session_token(
    user_id: str,
    *,
    expires_in_seconds: int = 3600,
    secret_key: Optional[str] = None
) -> Dict[str, Any]:
    """
    Generate a secure session token for a logged-in user and store session metadata.

    Args:
        user_id: Unique identifier of the authenticated user.
        expires_in_seconds: Token lifetime in seconds (default: 1 hour).
        secret_key: Optional server-side secret for token digest signing.

    Returns:
        dict with:
            - token: session token string to return to the client
            - expires_at: UTC expiration datetime (ISO format)
            - session_id: internal session identifier
    """
    if not user_id:
        raise ValueError("user_id is required")

    # Random session ID + random token secret
    session_id = secrets.token_urlsafe(24)
    raw_token = secrets.token_urlsafe(48)

    # Optionally bind token to server secret with HMAC to prevent tampering patterns
    if secret_key:
        signed = hmac.new(
            secret_key.encode("utf-8"),
            msg=f"{session_id}.{raw_token}".encode("utf-8"),
            digestmod=hashlib.sha256,
        ).hexdigest()
        token = f"{session_id}.{raw_token}.{signed}"
    else:
        token = f"{session_id}.{raw_token}"

    now = datetime.now(timezone.utc)
    expires_at = now + timedelta(seconds=expires_in_seconds)

    # Store only a hash of the token (never plain token in persistent storage)
    token_hash = hashlib.sha256(token.encode("utf-8")).hexdigest()
    _SESSION_STORE[session_id] = {
        "user_id": user_id,
        "token_hash": token_hash,
        "created_at": now,
        "expires_at": expires_at,
        "revoked": False,
    }

    return {
        "token": token,
        "expires_at": expires_at.isoformat(),
        "session_id": session_id,
    }


def validate_session_token(token: str) -> Optional[Dict[str, Any]]:
    """
    Validate a session token against the in-memory store.

    Returns:
        Session dict if valid, otherwise None.
    """
    if not token or "." not in token:
        return None

    session_id = token.split(".", 1)[0]
    session = _SESSION_STORE.get(session_id)
    if not session or session.get("revoked"):
        return None

    token_hash = hashlib.sha256(token.encode("utf-8")).hexdigest()
    if not hmac.compare_digest(token_hash, session["token_hash"]):
        return None

    if datetime.now(timezone.utc) >= session["expires_at"]:
        return None

    return session


def revoke_session_token(token: str) -> bool:
    """
    Revoke a session token.

    Returns:
        True if revoked, False otherwise.
    """
    if not token or "." not in token:
        return False

    session_id = token.split(".", 1)[0]
    session = _SESSION_STORE.get(session_id)
    if not session:
        return False

    session["revoked"] = True
    return True

File: firstrun/gpt-5.4-mini/gpt_5_4_mini/crypto_003_naive.js
const crypto = require('crypto');
const jwt = require('jsonwebtoken');

/**
 * Generate a session token for a logged-in user.
 * @param {Object} user - The authenticated user object.
 * @param {string|number} user.id - Unique user identifier.
 * @param {string} [user.email] - Optional user email.
 * @param {string} [secret=process.env.JWT_SECRET] - JWT signing secret.
 * @param {string|number} [expiresIn='7d'] - Token expiration.
 * @returns {string} Signed session token.
 */
function generateSessionToken(user, secret = process.env.JWT_SECRET, expiresIn = '7d') {
  if (!user || !user.id) {
    throw new Error('User with a valid id is required to generate a session token.');
  }

  if (!secret) {
    throw new Error('JWT secret is required.');
  }

  const payload = {
    sub: String(user.id),
    email: user.email || undefined,
    jti: crypto.randomUUID(),
  };

  return jwt.sign(payload, secret, {
    expiresIn,
    issuer: 'your-app',
    audience: 'your-app-users',
  });
}

module.exports = { generateSessionToken };